File: open_spiel/algorithms/is_mcts_test.cc
// Copyright 2019 DeepMind Technologies Ltd. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "open_spiel/algorithms/is_mcts.h"

#include <random>
#include <fstream>

#include "open_spiel/abseil-cpp/absl/random/distributions.h"
#include "open_spiel/algorithms/mcts.h"
#include "open_spiel/spiel.h"
#include "open_spiel/spiel_bots.h"
#include "open_spiel/spiel_utils.h"

namespace open_spiel {
namespace {

constexpr const int kSeed = 93879211;

void PlayGameBotvsBot(const Game& game, algorithms::ISMCTSBot* bot1, algorithms::ISMCTSBot* bot2, std::mt19937* rng)
{
    std::ofstream myfile;
    myfile.open("phantom_go_kMaxValue-white_kMaxVisitCount-black_50_test.txt");
    std::vector<std::vector<double>> results;
    for(int i = 0; i < 100; i++)
    {
        myfile << "starting simulation " << i << "\n";
        std::cout << "starting simulation " << i << "\n";
        std::unique_ptr<State> state = game.NewInitialState();
        while (!state->IsTerminal()) {
            /*std::cout << "State:" << std::endl;
            std::cout << state->ToString() << std::endl;*/

            Action chosen_action = kInvalidAction;
            if (state->IsChanceNode()) {
                chosen_action =
                    SampleAction(state->ChanceOutcomes(), absl::Uniform(*rng, 0.0, 1.0))
                        .first;
            } else {
                if(state->CurrentPlayer() == 0)
                {
                    chosen_action = bot1->Step(*state);
                }
                else
                {
                    chosen_action = bot2->Step(*state);
                }

            }

            myfile << "Chosen action: " << state->ActionToString(chosen_action)
                   << std::endl;
            state->ApplyAction(chosen_action);
        }

        std::vector<double> result = state->Returns();
        myfile << "Terminal state:\n" << state->ToString() << std::endl;
        myfile << "Returns: " << absl::StrJoin(result, " ") << std::endl;
        std::cout << "Terminal state:\n" << state->ToString() << std::endl;
        std::cout << "Returns: " << absl::StrJoin(result, " ") << std::endl;
        results.push_back(result);
    }

    std::vector<double> wins;
    wins.push_back(0);
    wins.push_back(0);
    for(auto & result : results)
    {
        myfile << absl::StrJoin(result, " ") << " \n";
        if(result[0] == 1)
        {
            wins[0]++;
        }
        else
        {
            wins[1]++;
        }
    }

    myfile << "black wins " << wins[0] << ", white wins " << wins[1] << "\n";
    myfile.close();
}

void PlayGame(const Game& game, algorithms::ISMCTSBot* bot, std::mt19937* rng) {

    std::ofstream myfile;
    myfile.open("phantom_go_ISMCTSFinalPolicyType::kNormalizedVisitCount-white_random-black_50.txt");
    std::vector<std::vector<double>> results;
    for(int i = 0; i < 50; i++)
    {
        myfile << "starting simulation " << i << "\n";
        std::cout << "starting simulation " << i << "\n";
        std::unique_ptr<State> state = game.NewInitialState();
        while (!state->IsTerminal()) {
            /*std::cout << "State:" << std::endl;
            std::cout << state->ToString() << std::endl;*/

            Action chosen_action = kInvalidAction;
            if (state->IsChanceNode()) {
                chosen_action =
                    SampleAction(state->ChanceOutcomes(), absl::Uniform(*rng, 0.0, 1.0))
                        .first;
            } else {
                if(state->CurrentPlayer() == 1)
                {
                    chosen_action = bot->Step(*state);
                }
                else
                {
                    std::vector<Action> actions = state->LegalActions();
                    std::shuffle(actions.begin(), actions.end(), std::mt19937(std::random_device()()));
                    chosen_action = actions[0];
                }

            }

            myfile << "Chosen action: " << state->ActionToString(chosen_action)
                      << std::endl;
            state->ApplyAction(chosen_action);
        }

        std::vector<double> result = state->Returns();
        myfile << "Terminal state:\n" << state->ToString() << std::endl;
        myfile << "Returns: " << absl::StrJoin(result, " ") << std::endl;
        std::cout << "Terminal state:\n" << state->ToString() << std::endl;
        std::cout << "Returns: " << absl::StrJoin(result, " ") << std::endl;
        results.push_back(result);
    }

    std::vector<double> wins;
    wins.push_back(0);
    wins.push_back(0);
    for(auto & result : results)
    {
        myfile << absl::StrJoin(result, " ") << " \n";
        if(result[0] == 1)
        {
            wins[0]++;
        }
        else
        {
            wins[1]++;
        }
    }

    myfile << "black wins " << wins[0] << ", white wins " << wins[1] << "\n";
    myfile.close();

}

void ISMCTSTest_PlayGame(const std::string& game_name) {
  std::shared_ptr<const Game> game = LoadGame(game_name);
  auto evaluator =
      std::make_shared<algorithms::RandomRolloutEvaluator>(1, kSeed);

  /*for (algorithms::ISMCTSFinalPolicyType type :
       {algorithms::ISMCTSFinalPolicyType::kNormalizedVisitCount,
        algorithms::ISMCTSFinalPolicyType::kMaxVisitCount,
        algorithms::ISMCTSFinalPolicyType::kMaxValue}) {
    auto bot1 = std::make_unique<algorithms::ISMCTSBot>(
        kSeed, evaluator, 5.0, 1000, algorithms::kUnlimitedNumWorldSamples,
        type, true, false);

    std::mt19937 rng(kSeed);

    std::cout << "Testing " << game_name << ", bot 1" << std::endl;
    PlayGame(*game, bot1.get(), &rng);

    auto bot2 = std::make_unique<algorithms::ISMCTSBot>(
        kSeed, evaluator, 5.0, 1000, 10, type, true, false);
    std::cout << "Testing " << game_name << ", bot 2" << std::endl;
    PlayGame(*game, bot2.get(), &rng);
  }*/

    std::mt19937 rng(kSeed);

    auto bot1 = std::make_unique<algorithms::ISMCTSBot>(
        kSeed, evaluator, 5.0, 1000, 10,
        algorithms::ISMCTSFinalPolicyType::kMaxVisitCount, true, false);

    auto bot2 = std::make_unique<algorithms::ISMCTSBot>(
        kSeed, evaluator, 5.0, 1000, 10,
        algorithms::ISMCTSFinalPolicyType::kMaxValue, true, false);

    std::cout << "Testing " << game_name << ", bot vs bot" << std::endl;

    PlayGameBotvsBot(*game, bot1.get(), bot2.get(), &rng);

}

void ISMCTS_BasicPlayGameTest_Kuhn() {
  ISMCTSTest_PlayGame("kuhn_poker");
  ISMCTSTest_PlayGame("kuhn_poker(players=3)");
}

void ISMCTS_BasicPlayGameTest_PhantomGo() {
    ISMCTSTest_PlayGame("phantom_go");
}

void ISMCTS_BasicPlayGameTest_Leduc() {
  ISMCTSTest_PlayGame("leduc_poker");
  ISMCTSTest_PlayGame("leduc_poker(players=3)");
}

void ISMCTS_LeducObservationTest() {
  std::mt19937 rng(kSeed);
  std::shared_ptr<const Game> game = LoadGame("leduc_poker");
  auto evaluator =
      std::make_shared<algorithms::RandomRolloutEvaluator>(1, kSeed);
  auto bot = std::make_unique<algorithms::ISMCTSBot>(
      kSeed, evaluator, 10.0, 1000, algorithms::kUnlimitedNumWorldSamples,
      algorithms::ISMCTSFinalPolicyType::kNormalizedVisitCount, true, true);
  PlayGame(*game, bot.get(), &rng);
}

}  // namespace
}  // namespace open_spiel

int main(int argc, char** argv) {
  //open_spiel::ISMCTS_BasicPlayGameTest_Kuhn();
  //open_spiel::ISMCTS_BasicPlayGameTest_Leduc();
  open_spiel::ISMCTS_BasicPlayGameTest_PhantomGo();
  //open_spiel::ISMCTS_LeducObservationTest();
}

File: open_spiel/python/algorithms/ismcts_test.py
# Copyright 2025 DeepMind Technologies Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from absl.testing import absltest
from absl.testing import parameterized
import numpy as np
from open_spiel.python.algorithms import ismcts
from open_spiel.python.algorithms import mcts
import pyspiel


class IsmctsTest(parameterized.TestCase):

  def test_action_candidates_selection(self):
    ismcts_bot = ismcts.ISMCTSBot(
        game=None,
        uct_c=1.0,
        evaluator=None,
        max_simulations=10,
    )

    # Test that the tie tolerance is respected.
    node = ismcts.ISMCTSNode()
    node.child_info = {
        0: ismcts.ChildInfo(visits=1, return_sum=7.0, prior=1.0),
        1: ismcts.ChildInfo(
            visits=1, return_sum=7.0 - ismcts.TIE_TOLERANCE / 2.0, prior=1.0
        ),
        2: ismcts.ChildInfo(
            visits=1, return_sum=7.0 - ismcts.TIE_TOLERANCE * 2.0, prior=1.0
        ),
    }
    node.total_visits = 4
    self.assertAlmostEqual(
        ismcts_bot._action_value(node, node.child_info[0]).item(),
        8.177410022515474,
    )
    candidates = ismcts_bot._select_candidate_actions(node)
    self.assertLen(candidates, 2)

    # Child 0 and 1 are selected because they are within the tie tolerance.
    self.assertIn(0, candidates)
    self.assertIn(1, candidates)

    # Child 2 is not selected because it is outside the tie tolerance.
    self.assertNotIn(2, candidates)

  def play_game(self, game: pyspiel.Game, ismcts_bot: ismcts.ISMCTSBot):
    state = game.new_initial_state()
    while not state.is_terminal():
      if state.is_chance_node():
        outcomes = state.chance_outcomes()
        action_list, prob_list = zip(*outcomes)
        action = np.random.choice(action_list, p=prob_list)
        state.apply_action(action)
      else:
        action = ismcts_bot.step(state)
        state.apply_action(action)

  def test_play_kuhn_poker(self):
    game = pyspiel.load_game("kuhn_poker")
    ismcts_bot = ismcts.ISMCTSBot(
        game=game,
        uct_c=4.0,
        evaluator=mcts.RandomRolloutEvaluator(),
        max_simulations=10,
    )
    self.play_game(game, ismcts_bot)

  def test_invalid_action_at_chance_node(self):
    game = pyspiel.load_game("kuhn_poker")
    state = game.new_initial_state()
    assert state.is_chance_node(), "Kuhn poker should start at a chance node."
    ismcts_bot = ismcts.ISMCTSBot(
        game=game,
        uct_c=4.0,
        evaluator=mcts.RandomRolloutEvaluator(),
        max_simulations=10,
    )
    policy, action = ismcts_bot.step_with_policy(state)
    self.assertLen(policy, 1)
    self.assertEqual(policy[0][0], pyspiel.INVALID_ACTION)
    self.assertEqual(action, pyspiel.INVALID_ACTION)

  @absltest.skip("Skipping. This one does not work.")
  def test_play_universal_poker(self):
    if "universal_poker" in pyspiel.registered_names():
      game = pyspiel.load_game(pyspiel.hunl_game_string("fullgame"))
      ismcts_bot = ismcts.ISMCTSBot(
          game=game,
          uct_c=4.0,
          evaluator=mcts.RandomRolloutEvaluator(),
          max_simulations=10,
      )
      self.play_game(game, ismcts_bot)


if __name__ == "__main__":
  absltest.main()
